power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp

```diff
@@ -495,17 +518,61 @@ class Serializer {
-        return ctype_func_selector(attribute.ctype, [element_ptr, &attribute]<class T> {
-            return is_nan(attribute.get_attribute<T const>(element_ptr));
+    static bool check_nan(BufferView const& element_buffer, MetaComponent const& component,
+                          MetaAttribute const& attribute) {
+        if (element_buffer.buffer->data != nullptr) {
+            RawElementPtr element_ptr = component.advance_ptr(element_buffer.buffer->data, element_buffer.idx);
+            return ctype_func_selector(attribute.ctype, [element_ptr, &attribute]<class T> {
+                return is_nan(attribute.get_attribute<T const>(element_ptr));
+            });
+        } else if (auto it = std::ranges::find_if(element_buffer.buffer->attributes,
+                                                  [&attribute](auto const& attribute_buffer) {
+                                                      return attribute_buffer.meta_attribute == &attribute;
+                                                  });
+                   it != element_buffer.buffer->attributes.end()) {
+            return check_nan(*it, element_buffer.idx);
+        }
+        return true;
+    }
+
+    static bool check_nan(AttributeBuffer<void const> const& attribute_buffer, Idx idx) {
+        return check_all_nan(attribute_buffer, idx, 1);
+    }
+
+    static bool check_all_nan(AttributeBuffer<void const> const& attribute_buffer, Idx idx, Idx size) {
+        return ctype_func_selector(attribute_buffer.meta_attribute->ctype, [&]<class T> {
+            return std::ranges::all_of(
+                std::span<T const>{reinterpret_cast<T const*>(attribute_buffer.data) + idx, static_cast<size_t>(size)},
+                [](auto const& x) { return is_nan(x); });
         });
     }
 
-    void pack_attribute(RawElementPtr element_ptr, MetaAttribute const& attribute) {
-        ctype_func_selector(attribute.ctype, [this, element_ptr, &attribute]<class T> {
-            packer_.pack(attribute.get_attribute<T const>(element_ptr));
+    void pack_attribute(BufferView const& element_buffer, MetaComponent const& component,
+                        MetaAttribute const& attribute) {
+        if (element_buffer.buffer->data != nullptr) {
+            RawElementPtr element_ptr = component.advance_ptr(element_buffer.buffer->data, element_buffer.idx);
+            ctype_func_selector(attribute.ctype, [this, element_ptr, &attribute]<class T> {
+                packer_.pack(attribute.get_attribute<T const>(element_ptr));
+            });
+        } else if (auto it = std::ranges::find_if(element_buffer.buffer->attributes,
+                                                  [&attribute](auto const& attribute_buffer) {
+                                                      return attribute_buffer.meta_attribute == &attribute;
+                                                  });
+                   it != element_buffer.buffer->attributes.end()) {
```

Comment:
this is a potentially expensive search in the hot loop. Unfortunately, I have tried many different things but haven't been able to come up with a satisfactory solution except for making 2 completely separate control flows (one for row-based, one for columnar).

@TonyXiang8787 what do you think?

Reply:
We can use similar strategy as in de-serializer. At the beginning we pre-map all the columnar buffers to the sequence number of all the attributes of a component. Then we do not do the name look-up afterwards.